### PR TITLE
refactor(x/participationrewards/keeper): combine GetProtocolData+UnmarshalProtocolData

### LINKED ISSUE #1631: Add GetAndUnmarshalProtocolData and MarshalAndSetProtocolData methods
Getting and Setting protocol data involves lots of repeated logic. We can simplify and bunch of code with functions that combine the two actions.

### PR DESCRIPTION
**Pull request by odeke-em**
This change combines GetProtocolData and types.UnmarshalProtocolData into a generic function that unifies the functionality and the pattern.

Fixes #1631

## Summary by CodeRabbit

- **Refactor**
  - Improved data retrieval and unmarshalling process for various pool types to enhance code maintainability and simplify error handling.
  
- **Bug Fixes**
  - Corrected data assignments in callback functions to ensure accurate protocol data updates and enhance system reliability.

### CODE REVIEW

**Review by joe-bowman — DISMISSED** [review]
looks great, thank you!

File: x/participationrewards/keeper/distribution.go [hunk 1]
```diff
@@ -37,16 +37,12 @@ func DepthFirstSearch(graph AssetGraph, visited map[string]struct{}, asset strin
 func (k *Keeper) CalcTokenValues(ctx sdk.Context) (TokenValues, error) {
 	k.Logger(ctx).Info("calcTokenValues")
 
-	data, found := k.GetProtocolData(ctx, types.ProtocolDataTypeOsmosisParams, "osmosisparams")
-	if !found {
-		return TokenValues{}, errors.New("could not find osmosisparams protocol data")
-	}
-	osmoParams, err := types.UnmarshalProtocolData(types.ProtocolDataTypeOsmosisParams, data.Data)
+	data, osmoParams, err := GetAndUnmarshalProtocolData[*types.OsmosisParamsProtocolData](ctx, "osmosisparams", types.ProtocolDataTypeOsmosisParams)
```
Comment (tropicaldog) [inline x/participationrewards/keeper/distribution.go:40]:
```suggestion
	data, osmoParams, err := GetAndUnmarshalProtocolData[*types.OsmosisParamsProtocolData](ctx, k "osmosisparams", types.ProtocolDataTypeOsmosisParams)
```

Comment (tropicaldog) [reply x/participationrewards/keeper/distribution.go:40]: Please add the missing keeper argument.

**Review by tropicaldog — CHANGES_REQUESTED** [review]

**Comment by tropicaldog** [comment]
Missing keeper argument also exists on several places, similar to the ‘distribution.go’ file, please fix @odeke-em

**Comment by odeke-em** [comment]
> Missing keeper argument also exists on several places, similar to the ‘distribution.go’ file, please fix @odeke-em 

Done, thanks for the ping @tropicaldog

**Review by tropicaldog — APPROVED** [review]
LGTM. failing ci is pending fix on #1680

**Review by tropicaldog — APPROVED** [review]
lgtm. Thanks @odeke-em!

**Comment by joe-bowman** [comment]
This pull request has been automatically marked as stale because it has not had any recent activity. It will be closed if no further activity occurs. Thank you!

**Review by joe-bowman — APPROVED** [review]
lgtm

**Pull request merged**